core/src/main/java/cz/o2/proxima/storage/watermark/GlobalWatermarkThroughputLimiter.java

```diff
@@ -0,0 +1,143 @@
+  private transient String processName = UUID.randomUUID().toString();
+
+  private long lastGlobalWatermarkUpdate = Long.MIN_VALUE;
+  private long globalWatermark = Long.MIN_VALUE;
+  private long sleepTimeMs = globalWatermarkUpdatePeriodMs;
+
+  @Override
+  public void setup(Map<String, Object> cfg) {
+    this.tracker =
+        Optional.ofNullable(cfg.get(TRACKER_CFG_PREFIX + KW_CLASS))
+            .map(Object::toString)
+            .map(
+                cls ->
+                    ExceptionUtils.uncheckedFactory(
+                        () -> Classpath.newInstance(cls, GlobalWatermarkTracker.class)))
+            .orElseThrow(
+                () ->
+                    new IllegalArgumentException(
+                        String.format(
+                            "Missing %s%s specifying GlobalWatermarkTracker implementation",
+                            TRACKER_CFG_PREFIX, KW_CLASS)));
+    this.tracker.setup(
+        cfg.entrySet()
+            .stream()
+            .filter(e -> e.getKey().startsWith(TRACKER_CFG_PREFIX))
+            .map(e -> Pair.of(e.getKey().substring(TRACKER_CFG_PREFIX.length()), e.getValue()))
+            .collect(Collectors.toMap(Pair::getFirst, Pair::getSecond)));
+    this.maxAheadTimeFromGlobalMs = getLong(cfg, MAX_AHEAD_TIME_MS_CFG, maxAheadTimeFromGlobalMs);
+    this.globalWatermarkUpdatePeriodMs =
+        getLong(cfg, GLOBAL_WATERMARK_UPDATE_MS_CFG, globalWatermarkUpdatePeriodMs);
+    this.sleepTimeMs = getLong(cfg, DEFAULT_SLEEP_TIME_CFG, sleepTimeMs);
+  }
+
+  private Long getLong(Map<String, Object> cfg, String key, long defVal) {
+    return Optional.ofNullable(cfg.get(key))
+        .map(Object::toString)
+        .map(Long::valueOf)
+        .orElse(defVal);
+  }
+
+  @Override
+  public Duration getPauseTime(Context context) {
+    updateGlobalWatermarkIfNeeded(context);
+    if (globalWatermark + maxAheadTimeFromGlobalMs < context.getMinWatermark()) {
+      log.info(
+          "ThroughputLimiter {} pausing processing for {} ms on global watermark {} and context.minWatermark {}",
+          this,
+          sleepTimeMs,
+          globalWatermark,
+          context.getMinWatermark());
+      return Duration.ofMillis(sleepTimeMs);
+    }
+    return Duration.ZERO;
+  }
+
+  @Override
+  public void close() {
+    if (tracker != null) {
+      ExceptionUtils.unchecked(tracker::close);
+    }
+  }
+
+  @Override
+  public String toString() {
+    return MoreObjects.toStringHelper(this)
+        .add("tracker", tracker)
+        .add("maxAheadTimeFromGlobalMs", maxAheadTimeFromGlobalMs)
+        .add("globalWatermarkUpdatePeriodMs", globalWatermarkUpdatePeriodMs)
+        .add("processName", processName)
+        .add("sleepTimeMs", sleepTimeMs)
+        .toString();
+  }
+
+  private void updateGlobalWatermarkIfNeeded(Context context) {
+    long now = System.currentTimeMillis();
+    if (now - globalWatermarkUpdatePeriodMs > lastGlobalWatermarkUpdate) {
```

Comment:
this code path may be tricky to test, is it covered?

Reply:
Do you think because of the `System.currentTimeMillis()`? It seems covered.

Reply:
yes